staging/src/k8s.io/cloud-provider/controllers/route/route_controller.go

```diff
@@ -79,16 +96,76 @@ func New(routes cloudprovider.Routes, kubeClient clientset.Interface, nodeInform
 		kubeClient:       kubeClient,
 		clusterName:      clusterName,
 		clusterCIDRs:     clusterCIDRs,
+		nodeInformer:     nodeInformer,
 		nodeLister:       nodeInformer.Lister(),
 		nodeListerSynced: nodeInformer.Informer().HasSynced,
 	}
 
-	return rc
+	if utilfeature.DefaultFeatureGate.Enabled(features.CloudControllerManagerWatchBasedRoutesReconciliation) {
+		// We use [TypedWithMaxWaitRateLimiter] to cap the [TypedBucketRateLimiter] at 10 seconds.
+		// Without this cap, the bucket rate limiter would keep accumulating delay for each node event.
+		// With the cap, even under a constant stream of node events, reconciles happen at most every
+		// 10 seconds — either triggered immediately or queued for another run if one is already in progress.
+		bucketRateLimiter := &workqueue.TypedBucketRateLimiter[string]{Limiter: rate.NewLimiter(rate.Every(minRouteResyncInterval), 1)}
+		rc.workqueue = workqueue.NewTypedRateLimitingQueueWithConfig(
+			workqueue.NewTypedWithMaxWaitRateLimiter[string](bucketRateLimiter, minRouteResyncInterval),
+			workqueue.TypedRateLimitingQueueConfig[string]{Name: "Routes"},
+		)
+
+		_, err := rc.nodeInformer.Informer().AddEventHandler(
+			// It is only necessary to reconcile the routes for any events that have the potential to impact them:
+			// - Node is added
+			// - Node is removed
+			// - Node .Status.Addresses is changed
+			// - Node .Spec.PodCIDRs is changed
+			cache.ResourceEventHandlerFuncs{
+				AddFunc:    rc.enqueueReconcile,
+				UpdateFunc: rc.handleNodeUpdate,
+				DeleteFunc: rc.enqueueReconcile,
+			},
+		)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return rc, nil
+}
+
+func (rc *RouteController) enqueueReconcile(_ interface{}) {
+	// reconcileNodeRoutes always reconciles the full cluster. It does not make sense to have
+	// separate entries in the workqueue per node, but a single one for the whole cluster is enough.
+	rc.workqueue.AddRateLimited("routes")
+}
+
+func (rc *RouteController) handleNodeUpdate(oldObj, newObj interface{}) {
+
+	oldNode, oldOk := oldObj.(*v1.Node)
+	newNode, newOk := newObj.(*v1.Node)
+
+	if !oldOk || !newOk {
+		return
+	}
+
+	// The Node informer triggers a periodic update event.
+	// In these cases, the old and new Node objects are identical. We use this event as a signal to perform
+	// a route reconciliation — our regular cleanup process — as described in the KEP 5237.
+	resync := oldNode.GetResourceVersion() == newNode.GetResourceVersion()
+	diffInPodCIDR := !reflect.DeepEqual(oldNode.Spec.PodCIDRs, newNode.Spec.PodCIDRs)
+	diffInNodeAddresses := !reflect.DeepEqual(oldNode.Status.Addresses, newNode.Status.Addresses)
+
+	if diffInPodCIDR || diffInNodeAddresses || resync {
+		rc.enqueueReconcile(newObj)
+	}
 }
```

Comment:
/hold

but this means that if you have 10k nodes, in each resync you are going to trigger 10k reconcilations

Reply:
ok, I got it , this is misleading, 

> func (rc *RouteController) enqueueReconcile(_ interface{}) {
	// reconcileNodeRoutes always reconciles the full cluster. It does not make sense to have
	// separate entries in the workqueue per node, but a single one for the whole cluster is enough.
	rc.workqueue.AddRateLimited("routes")
}

/hold cancel